src/utils/Field.ts

```diff
@@ -0,0 +1,311 @@
+ * myField.value[0] === 'something'    (repeating)
+ * myField.value === ['something','another']    (repeating - all fields)
+ *
+ * NOTE: I tried to make this a subclass of Array<Field> but that didn't work out.  Extending
+ * the builtin array type requires native true ES6 syntax and with webpack it seems to be
+ * not supported.  It would have been a much cleaner implementation and cleaner and simpler syntax
+ * for working with repeats.
+ * The syntax above is a bit more verbose but will have to do for now.
+ * SEE: https://stackoverflow.com/questions/35128241/extreme-es6-array-extension-weirdness-it-reverts-to-regular-array-with-webpack
+ */
+
+export default class Field {
+  _fieldKey = '';
+  _formUuid = '';
+  _repeatingFields = new Array<Field>();
+  _sourceField: any = null;
+
+  constructor(fieldKey: string, formUuid: string) {
+    this._fieldKey = fieldKey;
+    this._formUuid = formUuid;
+  }
+
+  get id(): string {
+    return this._fieldKey;
+  }
+
+  // raw field value
+  get value(): FeatheryFieldTypes {
+    // need to track changes via a proxy to any field value that is an Object or Array
+    if (
+      fieldValues[this._fieldKey] !== null &&
+      fieldValues[this._fieldKey] instanceof Object
+    )
+      return new Proxy(fieldValues[this._fieldKey] as object, {
+        set: (target: any, property: any, value) => {
+          target[property] = parseUserVal(value, this._fieldKey);
+          // we just need to know there is a change
+          updateFieldKeys.push(this._fieldKey);
+          notifyFieldsChanged();
+          return true;
+        }
+      });
+    return fieldValues[this._fieldKey];
+  }
+
+  // set raw field value
+  set value(val: FeatheryFieldTypes) {
+    if (Array.isArray(val))
+      fieldValues[this._fieldKey] = val.map((entry) =>
+        parseUserVal(entry, this._fieldKey)
+      );
+    else fieldValues[this._fieldKey] = parseUserVal(val, this._fieldKey);
+
+    // update
+    updateFieldKeys.push(this._fieldKey);
+    notifyFieldsChanged();
+  }
+
+  _getFormSpecificProps(): {
+    type: string;
+    displayText: string;
+    onThisForm: boolean;
+  } {
+    const field = this._getSourceField();
+    const type = field ? field.servar.type : '';
+    const displayText = field ? field.servar.name : '';
+    const onThisForm = !!field;
+    return { type, displayText, onThisForm };
+  }
+
+  _getSourceField(): any {
+    if (this._sourceField === null && internalState) {
+      this._sourceField = Object.values(
+        internalState[this._formUuid].steps
+      ).reduce((field: any, step: any) => {
+        if (field) return field;
+        return step.servar_fields.find(
+          (field: any) => field.servar.key === this._fieldKey
+        );
+      }, null);
+    }
+    return this._sourceField;
+  }
+
+  // read-only props that are only relevant to form fields
+  get type(): string {
+    const { type, onThisForm } = this._getFormSpecificProps();
+    if (!onThisForm)
+      console.warn(
```

Comment:
should these be console.error instead?

Reply:
Per our discussion, leaving it as warnings.  console.error is a bit too in your face.  It expands with a stack trace.  It could be a bit annoying for someone using console to debug a logic rule, etc.